java/vector/src/main/java/org/apache/arrow/vector/dictionary/StructSubfieldEncoder.java

```diff
@@ -0,0 +1,184 @@
+import org.apache.arrow.vector.util.TransferPair;
+
+/**
+ * Sub fields encoder/decoder for Dictionary encoded {@link StructVector}.
+ * Notes that child vectors within struct vector can either be dictionary encodeable or not.
+ */
+public class StructSubfieldEncoder {
+
+  private final BufferAllocator allocator;
+
+  private final DictionaryProvider.MapDictionaryProvider provider;
+  private final Map<Long, DictionaryHashTable> dictionaryIdToHashTable;
+
+  /**
+   * Construct an instance.
+   */
+  public StructSubfieldEncoder(
+      BufferAllocator allocator,
+      DictionaryProvider.MapDictionaryProvider provider) {
+
+    this.allocator = allocator;
+    this.provider = provider;
+
+    this.dictionaryIdToHashTable = new HashMap<>();
+
+    provider.getDictionaryIds().forEach(id ->
+        dictionaryIdToHashTable.put(id, new DictionaryHashTable(provider.lookup(id).getVector())));
+  }
+
+  private FieldVector getChildVector(StructVector vector, int index) {
+    return vector.getChildrenFromFields().get(index);
+  }
+
+  private StructVector cloneVector(StructVector vector) {
+
+    final FieldType fieldType = vector.getField().getFieldType();
+    StructVector cloned = (StructVector) fieldType.createNewSingleVector(
+        vector.getField().getName(), allocator, /*schemaCallback=*/null);
+
+    final ArrowFieldNode fieldNode = new ArrowFieldNode(vector.getValueCount(), vector.getNullCount());
+    cloned.loadFieldBuffers(fieldNode, vector.getFieldBuffers());
+
+    return cloned;
+  }
+
+  /**
+   * Dictionary encodes subfields for complex vector with a provided dictionary.
+   * The dictionary must contain all values in the sub fields vector.
+   * @param vector vector to encode
+   * @return dictionary encoded vector
+   */
+  public StructVector encode(StructVector vector, Map<Integer, Long> columnToDictionaryId) {
+    final int valueCount = vector.getValueCount();
+    final int childCount = vector.getChildrenFromFields().size();
+
+    List<Field> childrenFields = new ArrayList<>();
+
+    // initialize child fields.
+    for (int i = 0; i < childCount; i++) {
+      FieldVector childVector = getChildVector(vector, i);
+      Long dictionaryId = columnToDictionaryId.get(i);
+      // A null dictionaryId indicates the child vector shouldn't be encoded.
+      if (dictionaryId == null) {
+        childrenFields.add(childVector.getField());
+      } else {
+        Dictionary dictionary = provider.lookup(dictionaryId);
+        Preconditions.checkNotNull(dictionary, "Dictionary not found with id:" + dictionaryId);
+        FieldType indexFieldType = new FieldType(childVector.getField().isNullable(),
+            dictionary.getEncoding().getIndexType(), dictionary.getEncoding());
+        childrenFields.add(new Field(childVector.getField().getName(), indexFieldType, /*children=*/null));
+      }
+    }
+
+    // clone list vector and initialize data vector
+    StructVector encoded = cloneVector(vector);
```

Comment:
not for this PR, but it would be nice if we could consolidate child field subencoding in one place by generically manipulating ComplexVectors.

Reply:
Sure, I'll add it to my TODO list.
Could you give more description how to do this?

Reply:
I would need to think about it a little bit and review the APIs, probably it could be done by using the APIs on the fieldvector to get children vector but there might be some functionality missing.

Reply:
Ok, if you have any updates, could just open a JIRA and assign it to me.